File: apps/launcher/src-tauri/src/minecraft_files.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Component, Path, PathBuf};

const MAX_READ_BYTES: u64 = 2 * 1024 * 1024;
const OVERLAY_EXCLUDED_PREFIXES: &[&str] = &["saves/", "assets/", "logs/", "crash-reports/", "mods/"];

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MinecraftDirEntry {
    pub name: String,
    pub rel_path: String,
    pub is_dir: bool,
    pub has_persistent_override: bool,
    pub editable: bool,
}

pub fn persistent_minecraft_dir(inst_dir: &Path) -> PathBuf {
    inst_dir.join("persistent-minecraft")
}

pub fn minecraft_game_dir(inst_dir: &Path) -> PathBuf {
    inst_dir.join(".minecraft")
}

pub fn sanitize_minecraft_rel_path(path: &str) -> Result<PathBuf, String> {
    let trimmed = path.trim().replace('\\', "/");
    if trimmed.is_empty() {
        return Ok(PathBuf::new());
    }
    let mut out = PathBuf::new();
    for component in Path::new(&trimmed).components() {
        match component {
            Component::Normal(part) => out.push(part),
            Component::ParentDir => return Err("invalid path".into()),
            Component::RootDir | Component::Prefix(_) => return Err("invalid path".into()),
            Component::CurDir => {}
        }
    }
    Ok(out)
}

pub fn is_overlay_excluded(rel: &str) -> bool {
    let norm = rel.replace('\\', "/").to_lowercase();
    let norm = norm.strip_prefix("./").unwrap_or(&norm);
    OVERLAY_EXCLUDED_PREFIXES
        .iter()
        .any(|prefix| norm == prefix.trim_end_matches('/') || norm.starts_with(prefix))
}

pub fn is_path_editable(rel: &str) -> bool {
    if is_overlay_excluded(rel) {
        return false;
    }
    let lower = rel.to_lowercase();
    !lower.ends_with(".jar")
        && !lower.ends_with(".zip")
        && !lower.ends_with(".png")
        && !lower.ends_with(".jpg")
}

fn persistent_paths_set(inst_dir: &Path) -> Result<std::collections::HashSet<String>, String> {
    let mut set = std::collections::HashSet::new();
    let overlay = persistent_minecraft_dir(inst_dir);
    if !overlay.is_dir() {
        return Ok(set);
    }
    collect_rel_paths(&overlay, &overlay, &mut set)?;
    Ok(set)
}

fn collect_rel_paths(base: &Path, dir: &Path, out: &mut std::collections::HashSet<String>) -> Result<(), String> {
    for entry in fs::read_dir(dir).map_err(|e| e.to_string())? {
        let entry = entry.map_err(|e| e.to_string())?;
        let path = entry.path();
        let rel = path
            .strip_prefix(base)
            .map_err(|e| e.to_string())?
            .to_string_lossy()
            .replace('\\', "/");
        if path.is_dir() {
            collect_rel_paths(base, &path, out)?;
        } else {
            out.insert(rel);
        }
    }
    Ok(())
}

pub fn list_minecraft_entries(
    inst_dir: &Path,
    subpath: &str,
) -> Result<Vec<MinecraftDirEntry>, String> {
    let rel = sanitize_minecraft_rel_path(subpath)?;
    let mc = minecraft_game_dir(inst_dir);
    let dir = if rel.as_os_str().is_empty() {
        mc.clone()
    } else {
        mc.join(&rel)
    };
    if !dir.exists() {
        return Ok(vec![]);
    }
    if !dir.is_dir() {
        return Err("not a directory".into());
    }

    let persistent = persistent_paths_set(inst_dir)?;
    let mut entries = vec![];
    for entry in fs::read_dir(&dir).map_err(|e| e.to_string())? {
        let entry = entry.map_err(|e| e.to_string())?;
        let name = entry.file_name().to_string_lossy().to_string();
        let entry_rel = if rel.as_os_str().is_empty() {
            name.clone()
        } else {
            format!("{}/{}", rel.to_string_lossy(), name)
        };
        let norm_rel = entry_rel.replace('\\', "/");
        let is_dir = entry.path().is_dir();
        entries.push(MinecraftDirEntry {
            name,
            rel_path: norm_rel.clone(),
            is_dir,
            has_persistent_override: !is_dir && persistent.contains(&norm_rel),
            editable: !is_dir && is_path_editable(&norm_rel),
        });
    }
    entries.sort_by(|a, b| {
        b.is_dir
            .cmp(&a.is_dir)
            .then_with(|| a.name.to_lowercase().cmp(&b.name.to_lowercase()))
    });
    Ok(entries)
}

pub fn read_minecraft_file(inst_dir: &Path, rel_path: &str) -> Result<String, String> {
    let rel = sanitize_minecraft_rel_path(rel_path)?;
    if is_overlay_excluded(rel_path) {
        return Err("path is not readable".into());
    }
    if !is_path_editable(rel_path) {
        return Err("file is not editable".into());
    }
    let path = minecraft_game_dir(inst_dir).join(&rel);
    if !path.is_file() {
        return Err("file not found".into());
    }
    let meta = fs::metadata(&path).map_err(|e| e.to_string())?;
    if meta.len() > MAX_READ_BYTES {
        return Err("file too large to edit in launcher".into());
    }
    fs::read_to_string(&path).map_err(|e| e.to_string())
}

pub fn write_minecraft_file(
    inst_dir: &Path,
    rel_path: &str,
    content: &str,
    persist: bool,
) -> Result<(), String> {
    let rel = sanitize_minecraft_rel_path(rel_path)?;
    let norm = rel.to_string_lossy().replace('\\', "/");
    if is_overlay_excluded(&norm) {
        return Err("path is not writable".into());
    }
    if !is_path_editable(&norm) {
        return Err("file type is not editable".into());
    }

    let mc_path = minecraft_game_dir(inst_dir).join(&rel);
    if let Some(parent) = mc_path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    fs::write(&mc_path, content).map_err(|e| e.to_string())?;

    if persist {
        let overlay_path = persistent_minecraft_dir(inst_dir).join(&rel);
        if let Some(parent) = overlay_path.parent() {
            fs::create_dir_all(parent).map_err(|e| e.to_string())?;
        }
        fs::write(&overlay_path, content).map_err(|e| e.to_string())?;
    }
    Ok(())
}

pub fn delete_persistent_file(inst_dir: &Path, rel_path: &str) -> Result<(), String> {
    let rel = sanitize_minecraft_rel_path(rel_path)?;
    let overlay_path = persistent_minecraft_dir(inst_dir).join(&rel);
    if overlay_path.is_file() {
        fs::remove_file(&overlay_path).map_err(|e| e.to_string())?;
    }
    Ok(())
}

pub fn list_persistent_files(inst_dir: &Path) -> Result<Vec<String>, String> {
    let overlay = persistent_minecraft_dir(inst_dir);
    let mut set = std::collections::HashSet::new();
    if overlay.is_dir() {
        collect_rel_paths(&overlay, &overlay, &mut set)?;
    }
    let mut list: Vec<String> = set.into_iter().collect();
    list.sort();
    Ok(list)
}

pub fn apply_persistent_minecraft(inst_dir: &Path) -> Result<(), String> {
    let overlay = persistent_minecraft_dir(inst_dir);
    if !overlay.is_dir() {
        return Ok(());
    }
    let mc = minecraft_game_dir(inst_dir);
    apply_overlay_tree(&overlay, &overlay, &mc)
}

fn apply_overlay_tree(base: &Path, dir: &Path, mc_root: &Path) -> Result<(), String> {
    for entry in fs::read_dir(dir).map_err(|e| e.to_string())? {
        let entry = entry.map_err(|e| e.to_string())?;
        let path = entry.path();
        let rel = path
            .strip_prefix(base)
            .map_err(|e| e.to_string())?
            .to_string_lossy()
            .replace('\\', "/");

        if path.is_dir() {
            if rel == "mods" {
                continue;
            }
            if is_overlay_excluded(&format!("{rel}/")) {
                continue;
            }
            apply_overlay_tree(base, &path, mc_root)?;
        } else {
            if is_overlay_excluded(&rel) {
                continue;
            }
            let dest = mc_root.join(&rel);
            if let Some(parent) = dest.parent() {
                fs::create_dir_all(parent).map_err(|e| e.to_string())?;
            }
            fs::copy(&path, &dest).map_err(|e| e.to_string())?;
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sanitize_rejects_parent_dir() {
        assert!(sanitize_minecraft_rel_path("../secret").is_err());
    }

    #[test]
    fn overlay_excludes_saves() {
        assert!(is_overlay_excluded("saves/world1/level.dat"));
        assert!(!is_overlay_excluded("config/gadomancy.cfg"));
    }
}

File: apps/launcher/electron/backend/minecraft-files.ts
import { watch, type FSWatcher } from "node:fs";
import fs from "node:fs/promises";
import path from "node:path";
import { exists, runConcurrent } from "./fs-utils";
import type { MinecraftDirEntry } from "./types";

const maxReadBytes = 2 * 1024 * 1024;
const excludedPrefixes = ["saves/", "assets/", "logs/", "crash-reports/", "mods/"];
const externalFileMirrors = new Map<
  string,
  {
    watcher: FSWatcher;
    timer: ReturnType<typeof setTimeout> | null;
    pending: Promise<void>;
    instance: string;
    sourcePath: string;
    relPath: string;
  }
>();

export function persistentMinecraftDir(instance: string): string {
  return path.join(instance, "persistent-minecraft");
}

export function minecraftGameDir(instance: string): string {
  return path.join(instance, ".minecraft");
}

export function sanitizeMinecraftRelPath(input: string): string {
  const trimmed = input.trim().replaceAll("\\", "/");
  if (!trimmed) return "";
  if (/^(?:\/+|[A-Za-z]:(?:\/|$))/u.test(trimmed) || trimmed.includes("\0")) throw new Error("invalid path");
  const parts = trimmed.split("/");
  if (parts.some((part) => part === ".." || part.includes("\0"))) throw new Error("invalid path");
  return parts.filter((part) => part && part !== ".").join("/");
}

export function isOverlayExcluded(input: string): boolean {
  const normalized = input.replaceAll("\\", "/").toLowerCase().replace(/^\.\//, "");
  return excludedPrefixes.some((prefix) => normalized === prefix.slice(0, -1) || normalized.startsWith(prefix));
}

export function isPathEditable(input: string): boolean {
  if (isOverlayExcluded(input)) return false;
  const lower = input.toLowerCase();
  return ![".jar", ".zip", ".png", ".jpg"].some((extension) => lower.endsWith(extension));
}

export async function resolveMinecraftFilePath(instance: string, relPath: string): Promise<string> {
  const rel = sanitizeMinecraftRelPath(relPath);
  if (!rel || isOverlayExcluded(rel) || !isPathEditable(rel)) throw new Error("file type is not editable");
  if (/\.(?:exe|com|bat|cmd|msi|msp|scr|ps1|sh|app|lnk|url|desktop|dll|so|dylib)$/iu.test(rel)) {
    throw new Error("executable files cannot be opened from the file editor");
  }
  const gameRoot = await fs.realpath(minecraftGameDir(instance));
  const target = path.resolve(gameRoot, rel);
  const actual = await fs.realpath(target).catch(() => null);
  if (!actual || (actual !== gameRoot && !actual.startsWith(`${gameRoot}${path.sep}`))) throw new Error("file path is outside the Minecraft directory");
  const stat = await fs.stat(actual).catch(() => null);
  if (!stat?.isFile()) throw new Error("file not found");
  return actual;
}

function externalMirrorKey(instance: string, relPath: string): string {
  return path.resolve(minecraftGameDir(instance), sanitizeMinecraftRelPath(relPath));
}

async function mirrorExternalMinecraftFile(instance: string, sourcePath: string, relPath: string): Promise<void> {
  const gameRoot = await fs.realpath(minecraftGameDir(instance));
  const actualSource = await fs.realpath(sourcePath).catch(() => null);
  if (!actualSource || (actualSource !== gameRoot && !actualSource.startsWith(`${gameRoot}${path.sep}`))) return;
  const stat = await fs.stat(actualSource).catch(() => null);
  if (!stat?.isFile()) return;

  const overlayRoot = persistentMinecraftDir(instance);
  const overlayPath = path.resolve(overlayRoot, sanitizeMinecraftRelPath(relPath));
  await fs.mkdir(path.dirname(overlayPath), { recursive: true });
  const overlayParent = await fs.realpath(path.dirname(overlayPath));
  const instanceRoot = await fs.realpath(instance);
  if (overlayParent !== instanceRoot && !overlayParent.startsWith(`${instanceRoot}${path.sep}`)) return;
  const existing = await fs.lstat(overlayPath).catch(() => null);
  if (existing?.isSymbolicLink()) return;
  await fs.copyFile(actualSource, overlayPath);
}

export async function watchExternalMinecraftFile(instance: string, relPath: string): Promise<void> {
  const sourcePath = await resolveMinecraftFilePath(instance, relPath);
  const key = externalMirrorKey(instance, relPath);
  if (externalFileMirrors.has(key)) return;

  const watcher = watch(path.dirname(sourcePath), { persistent: false }, (_event, filename) => {
    if (filename != null && filename.toString() !== path.basename(sourcePath)) return;
    const entry = externalFileMirrors.get(key);
    if (!entry) return;
    if (entry.timer) clearTimeout(entry.timer);
    entry.timer = setTimeout(() => {
      entry.timer = null;
      entry.pending = entry.pending.then(() => mirrorExternalMinecraftFile(instance, sourcePath, relPath)).catch(() => undefined);
    }, 250);
  });
  watcher.on("error", () => void stopExternalMinecraftFileWatch(instance, relPath));
  externalFileMirrors.set(key, { watcher, timer: null, pending: Promise.resolve(), instance, sourcePath, relPath });
}

export async function stopExternalMinecraftFileWatch(instance: string, relPath: string): Promise<void> {
  const key = externalMirrorKey(instance, relPath);
  const entry = externalFileMirrors.get(key);
  if (!entry) return;
  if (entry.timer) clearTimeout(entry.timer);
  entry.watcher.close();
  externalFileMirrors.delete(key);
  await entry.pending;
}

export async function stopAllExternalMinecraftFileWatches(): Promise<void> {
  const pending: Promise<void>[] = [];
  for (const entry of externalFileMirrors.values()) {
    if (entry.timer) clearTimeout(entry.timer);
    entry.watcher.close();
    entry.pending = entry.pending.then(() => mirrorExternalMinecraftFile(entry.instance, entry.sourcePath, entry.relPath)).catch(() => undefined);
    pending.push(entry.pending);
  }
  externalFileMirrors.clear();
  await Promise.all(pending);
}

async function persistentPaths(instance: string): Promise<Set<string>> {
  const overlay = persistentMinecraftDir(instance);
  const result = new Set<string>();
  if (!(await exists(overlay))) return result;
  await runConcurrent([overlay], async (directory, enqueue) => {
    const entries = await fs.readdir(directory, { withFileTypes: true });
    for (const entry of entries) {
      const child = path.join(directory, entry.name);
      const rel = path.relative(overlay, child).replaceAll("\\", "/");
      if (entry.isDirectory()) enqueue(child);
      else result.add(rel);
    }
  });
  return result;
}

export async function listMinecraftEntries(instance: string, subpath: string): Promise<MinecraftDirEntry[]> {
  const rel = sanitizeMinecraftRelPath(subpath);
  const dir = path.join(minecraftGameDir(instance), rel);
  if (!(await exists(dir))) return [];
  const stat = await fs.stat(dir);
  if (!stat.isDirectory()) throw new Error("not a directory");
  const persistent = await persistentPaths(instance);
  const entries: MinecraftDirEntry[] = await Promise.all(
    (await fs.readdir(dir, { withFileTypes: true })).map(async (entry) => {
      const entryRel = rel ? `${rel}/${entry.name}` : entry.name;
      const isDir = entry.isDirectory();
      const editable = !isDir && isPathEditable(entryRel);
      const stat = editable ? await fs.stat(path.join(dir, entry.name)).catch(() => null) : null;
      return {
        name: entry.name,
        rel_path: entryRel,
        is_dir: isDir,
        has_persistent_override: !isDir && persistent.has(entryRel),
        editable,
        too_large_to_edit: Boolean(stat?.isFile() && stat.size > maxReadBytes),
      };
    }),
  );
  return entries.sort((a, b) => Number(b.is_dir) - Number(a.is_dir) || a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
}

export async function readMinecraftFile(instance: string, relPath: string): Promise<string> {
  const rel = sanitizeMinecraftRelPath(relPath);
  if (isOverlayExcluded(rel)) throw new Error("path is not readable");
  if (!isPathEditable(rel)) throw new Error("file is not editable");
  const target = path.join(minecraftGameDir(instance), rel);
  const stat = await fs.stat(target).catch(() => null);
  if (!stat?.isFile()) throw new Error("file not found");
  if (stat.size > maxReadBytes) throw new Error("file too large to edit in launcher");
  return fs.readFile(target, "utf8");
}

export async function writeMinecraftFile(instance: string, relPath: string, content: string, persist: boolean): Promise<void> {
  const rel = sanitizeMinecraftRelPath(relPath);
  if (isOverlayExcluded(rel) || !isPathEditable(rel)) throw new Error("file type is not editable");
  const target = path.join(minecraftGameDir(instance), rel);
  await fs.mkdir(path.dirname(target), { recursive: true });
  await fs.writeFile(target, content, "utf8");
  if (persist) {
    const overlay = path.join(persistentMinecraftDir(instance), rel);
    await fs.mkdir(path.dirname(overlay), { recursive: true });
    await fs.writeFile(overlay, content, "utf8");
  }
}

export async function deletePersistentFile(instance: string, relPath: string): Promise<void> {
  const rel = sanitizeMinecraftRelPath(relPath);
  await stopExternalMinecraftFileWatch(instance, rel);
  await fs.rm(path.join(persistentMinecraftDir(instance), rel), { force: true });
}

export async function listPersistentFiles(instance: string): Promise<string[]> {
  return [...(await persistentPaths(instance))].sort();
}

export async function applyPersistentMinecraft(instance: string): Promise<void> {
  const overlay = persistentMinecraftDir(instance);
  const targetRoot = minecraftGameDir(instance);
  async function visit(dir: string): Promise<void> {
    for (const entry of await fs.readdir(dir, { withFileTypes: true })) {
      const child = path.join(dir, entry.name);
      const rel = path.relative(overlay, child).replaceAll("\\", "/");
      if (entry.isDirectory()) {
        if (rel === "mods" || isOverlayExcluded(`${rel}/`)) continue;
        await visit(child);
      } else if (!isOverlayExcluded(rel)) {
        const target = path.join(targetRoot, rel);
        await fs.mkdir(path.dirname(target), { recursive: true });
        await fs.copyFile(child, target);
      }
    }
  }
  if (await exists(overlay)) await visit(overlay);
}
